samples/RefreshFlow/AngularApp/src/core/auth.service.ts

```diff
@@ -0,0 +1,164 @@
+        this.state$ = this.state.asObservable();
+
+        this.tokens$ = this.state.filter(state => state.authReady)
+            .map(state => state.tokens);
+
+        this.profile$ = this.state.filter(state => state.authReady)
+            .map(state => state.profile);
+
+        this.loggedIn$ = this.tokens$.map(tokens => !!tokens);
+    }
+
+    init(): Observable<AuthTokenModel> {
+        return this.startupTokenRefresh()
+            .do(() => this.scheduleRefresh());
+    }
+
+    register(data: RegisterModel): Observable<Response> {
+        return this.http.post('http://localhost:5056/account/register', data)
+            .catch(res => Observable.throw(res.json()));
+    }
+
+    login(user: LoginModel): Observable<any> {
+        return this.getTokens(user, 'password')
+            .catch(res => Observable.throw(res.json()))
+            .do(res => this.scheduleRefresh());
+    }
+
+    logout(): void {
+        this.updateState({ profile: null, tokens: null });
+        if (this.refreshSubscription$) {
+            this.refreshSubscription$.unsubscribe();
+        }
+        this.removeToken();
+    }
+
+    isInRole(usersRole: string): Observable<boolean> {
+        return this.profile$
+            .map(profile => profile.role.find(role => role === usersRole) !== undefined);
+    }
+
+    refreshTokens(): Observable<AuthTokenModel> {
+        return this.state.first()
+            .map(state => state.tokens)
+            .flatMap(tokens => this.getTokens({ refresh_token: tokens.refresh_token }, 'refresh_token')
+                .catch(error => Observable.throw('Session Expired'))
+            );
+    }
+
+    private storeToken(tokens: AuthTokenModel): void {
+        localStorage.setItem('auth-tokens', JSON.stringify(tokens));
+    }
+
+    private retrieveTokens(): AuthTokenModel {
+        const tokensString = localStorage.getItem('auth-tokens');
+        const tokensModel: AuthTokenModel = tokensString == null ? null : JSON.parse(tokensString);
+        return tokensModel;
+    }
+
+    private removeToken(): void {
+        localStorage.removeItem('auth-tokens');
+    }
+
+    private updateState(newState: AuthStateModel): void {
+        const previoudState = this.state.getValue();
+        this.state.next(Object.assign({}, previoudState, newState));
+    }
+
+    private getTokens(data: RefreshGrantModel | LoginModel, grantType: string): Observable<Response> {
+        const headers = new Headers({ 'Content-Type': 'application/x-www-form-urlencoded' });
+        const options = new RequestOptions({ headers: headers });
+
+        Object.assign(data, { grant_type: grantType, scope: 'openid offline_access' });
+
+        const params = new URLSearchParams();
+        Object.keys(data).forEach(key => params.append(key, (<any>data)[key]));
+
+        return this.http.post('http://localhost:5056/connect/token', params.toString(), options)
+            .map(res => {
```

Comment:
It seems on to be using `map` here. The callback is causing two side effects and then returning the original response, whereas a use of `map` usually expresses the intent to create a new array.